toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -0,0 +1,158 @@
+ public:
+  explicit LexerBenchHelper(llvm::StringRef text)
+      : source_(MakeSourceBuffer(text)) {}
+
+  auto Lex() -> TokenizedBuffer {
+    DiagnosticConsumer& consumer = NullDiagnosticConsumer();
+    return TokenizedBuffer::Lex(source_, consumer);
+  }
+
+  auto DiagnoseErrors() -> std::string {
+    std::string result;
+    llvm::raw_string_ostream out(result);
+    StreamDiagnosticConsumer consumer(out);
+    auto buffer = TokenizedBuffer::Lex(source_, consumer);
+    consumer.Flush();
+    CARBON_CHECK(buffer.has_errors())
+        << "Asked to diagnose errors but none found!";
+    return result;
+  }
+
+ private:
+  auto MakeSourceBuffer(llvm::StringRef text) -> SourceBuffer {
+    CARBON_CHECK(fs_.addFile(filename_, /*ModificationTime=*/0,
+                             llvm::MemoryBuffer::getMemBuffer(text)));
+    return std::move(*SourceBuffer::CreateFromFile(fs_, filename_));
+  }
+
+  llvm::vfs::InMemoryFileSystem fs_;
+  std::string filename_ = "test.carbon";
+  SourceBuffer source_;
+};
+
+// A large value for measurement stability without making benchmarking too slow.
+constexpr int NumTokens = 100000;
+
+void BM_ValidKeywords(benchmark::State& state) {
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i : llvm::seq(0, NumTokens)) {
+    static_cast<void>(i);
+    int token = absl::Uniform<int>(gen, 0, TokenKind::KeywordTokens.size());
+    os << sep << TokenKind::KeywordTokens[token].fixed_spelling();
+  }
+
+  LexerBenchHelper helper(source);
+  for (auto _ : state) {
+    TokenizedBuffer buffer = helper.Lex();
+    CARBON_CHECK(!buffer.has_errors());
+  }
+
+  state.counters["TokenRate"] = benchmark::Counter(
+      NumTokens, benchmark::Counter::kIsIterationInvariantRate);
+}
+BENCHMARK(BM_ValidKeywords);
+
+auto IdentifierStartChars() -> llvm::ArrayRef<char> {
+  static llvm::SmallVector<char> chars = [] {
+    llvm::SmallVector<char> chars;
+    chars.push_back('_');
+    for (char c : llvm::seq_inclusive('A', 'Z')) {
+      chars.push_back(c);
+    }
+    for (char c : llvm::seq_inclusive('a', 'z')) {
+      chars.push_back(c);
+    }
+    return chars;
+  }();
+  return chars;
+}
+
+auto IdentifierChars() -> llvm::ArrayRef<char> {
+  static llvm::SmallVector<char> chars = [] {
+    llvm::ArrayRef<char> start_chars = IdentifierStartChars();
+    llvm::SmallVector<char> chars(start_chars.begin(), start_chars.end());
+    for (char c : llvm::seq_inclusive('0', '9')) {
+      chars.push_back(c);
+    }
+    return chars;
+  }();
+  return chars;
+}
+
+void BM_ValidIdentifiers(benchmark::State& state) {
+  int min_length = state.range(0);
+  int max_length = state.range(1);
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i : llvm::seq(0, NumTokens)) {
+    static_cast<void>(i);
+    int length = absl::Uniform<int>(gen, min_length, max_length);
```

Comment:
Should this be:
```suggestion
    int length = absl::Uniform<int>(absl::IntervalClosedClosed, gen, min_length, max_length);
```
... to include the `length == max_length` case, or did you intend for a semi-open interval here? The naming (`max_length`) to me suggests a closed interval.

Reply:
I was happy with semi-open, but I also don't feel strongly so happy to make it closed.